src/js_subscriptions.hpp

```diff
@@ -0,0 +1,885 @@
+ *     argument, and which updates the SubscriptionSet as required
+ *   Argument 2: A callback to be called when the state of the SubscriptionSet is "Complete"
+ *      or "Error" after the update has been applied (see `wait_for_synchronization_impl`).
+ * @param return_value \ref Returns the return value of the update callback
+ *
+ * TODO handle async callbacks
+ */
+template <typename T>
+void SubscriptionSetClass<T>::update(ContextType ctx, ObjectType this_object, Arguments& args,
+                                     ReturnValue& return_value)
+{
+    args.validate_count(2);
+
+    FunctionType update_callback = Value::validated_to_function(ctx, args[0], "update callback");
+    FunctionType completion_callback = Value::validated_to_function(ctx, args[1], "completion callback");
+
+    Protected<FunctionType> protected_update_callback(ctx, update_callback);
+    Protected<FunctionType> protected_completion_callback(ctx, completion_callback);
+
+    Protected<ObjectType> protected_this(ctx, this_object);
+    Protected<typename T::GlobalContext> protected_ctx(js::Context<T>::get_global_context(ctx));
+
+    auto subs = get_internal<T, SubscriptionSetClass<T>>(ctx, this_object);
+
+    try {
+        HANDLESCOPE(protected_ctx);
+
+        // Create a mutable copy of this instance (which copies the original and upgrades
+        // its internal transaction to a write transaction, so we can make updates to it -
+        // SubscriptionSets are otherwise immutable)
+        auto mutable_subs_js = MutableSubscriptionSetClass<T>::create_instance(ctx, subs->make_mutable_copy());
+        auto mutable_subs = get_internal<T, MutableSubscriptionSetClass<T>>(ctx, mutable_subs_js);
+
+        // Call the provided callback, passing in the mutable copy as an argument,
+        // and return its return value
+        ValueType arguments[]{mutable_subs_js};
+        auto const& callback_return =
+            Function<T>::callback(protected_ctx, protected_update_callback, protected_this, 1, arguments);
+
+        // Commit the mutation, which downgrades its internal transaction to a read transaction
+        // so no more changes can be made to it, and returns a new (immutable) SubscriptionSet
+        // with the changes we made
+        auto new_sub_set = std::move(*mutable_subs).commit();
+
+        // Update this SubscriptionSetClass instance to point to the updated version
+        set_internal<T, SubscriptionSetClass<T>>(ctx, this_object,
+                                                 new SubscriptionSet<T>(std::move(new_sub_set), subs->sync_session));
+
+        // Asynchronously wait for the SubscriptionSet to be synchronised
+        SubscriptionSetClass<T>::wait_for_synchronization_impl(protected_ctx, protected_this,
+                                                               protected_completion_callback);
+    }
+    catch (...) {
+        throw;
+    }
```

Comment:
What does this add?

Reply:
Hm I suspect there might have been more than just this line in this handler at some point. Agreed it adds nothing, will remove!